Expose strict mode functionality

lukechilds opened the pull request:
This PR adds support for using `assert` in strict mode.

Resolves: #37 
Makes some progress towards: https://github.com/browserify/commonjs-assert/issues/32 

Ported from this Node.js PR: https://github.com/nodejs/node/pull/17002

Tests are all failing. This isn't due to any changes in this PR they are already failing in master: https://github.com/browserify/commonjs-assert/issues/33

I've tested locally on a couple of different Node.js versions and tests are passing but would feel a lot more confident getting this properly tested against all browser targets.

Happy to look into the Travis issue and submit a separate PR resolving that if you want.

goto-bus-stop commented on assert.js:
```diff
@@ -483,6 +483,18 @@ assert.doesNotThrow = function(block, /*optional*/error, /*optional*/message) {
+function strict(value, message) {
+  if (!value) fail(value, true, message, '==', strict);
+}
+assert.strict = Object.assign(strict, assert, {
```
This will need a ponyfill to maintain support for the very old nodes and browsers that we target right now

lukechilds replied:
Is it ok to depend on [this module](https://github.com/sindresorhus/object-assign) and then `require` it?

I noticed other shims (Buffer/Object.keys) are dumped inline rather than added as dependencies. Is there a reason for that?

lukechilds commented:
@goto-bus-stop addresses feedback, let me know if `require`ing the ponyfill is ok.

Also resolved some of the issues with the tests here: https://github.com/browserify/commonjs-assert/pull/42

goto-bus-stop commented:
I think the buffer shim is inlined because recent buffer versions don't work at all on old IE (throws an error on load). object-assign _should_ be ok, it seems like `assert` only supported IE9+ already… i'll have a closer look later to confirm what the current oldest supported version is and document it somewhere :P

lukechilds commented:
>I think the buffer shim is inlined because recent buffer versions don't work at all on old IE

Ahhhh, that makes sense.

>i'll have a closer look later to confirm what the current oldest supported version is and document it somewhere :P

That would be great, thanks!

lukechilds commented:
@goto-bus-stop is this ok to be merged?

lukechilds commented:
@goto-bus-stop should we publish a new release with this?

I don't have `npm` permissions.

goto-bus-stop commented:
ideally yes! i don't have publish perms either :'

@defunctzombie do you have a moment to give the browserify org members publish access on npm?
```
npm access grant read-write browserify:developers assert
```

lukechilds commented:
(or @calvinmetcalf or @solderjs 🙏)

lukechilds commented:
Any update on this?

Ping @defunctzombie @calvinmetcalf @solderjs
